File: C/red-black tree/meta-llama-llama-4-maverick-17b-128e-instruct.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Define the structure for a Red-Black Tree node
typedef struct Node {
    int data;
    int color; // 0 for red, 1 for black
    struct Node* left;
    struct Node* right;
    struct Node* parent;
} Node;

// Function to create a new node
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (!newNode) {
        printf("Memory error\n");
        return NULL;
    }
    newNode->data = data;
    newNode->color = 0; // New node is red
    newNode->left = newNode->right = newNode->parent = NULL;
    return newNode;
}

// Function to perform left rotation
void leftRotate(Node** root, Node* x) {
    Node* y = x->right;
    x->right = y->left;
    if (y->left != NULL) {
        y->left->parent = x;
    }
    y->parent = x->parent;
    if (x->parent == NULL) {
        *root = y;
    } else if (x == x->parent->left) {
        x->parent->left = y;
    } else {
        x->parent->right = y;
    }
    y->left = x;
    x->parent = y;
}

// Function to perform right rotation
void rightRotate(Node** root, Node* y) {
    Node* x = y->left;
    y->left = x->right;
    if (x->right != NULL) {
        x->right->parent = y;
    }
    x->parent = y->parent;
    if (y->parent == NULL) {
        *root = x;
    } else if (y == y->parent->right) {
        y->parent->right = x;
    } else {
        y->parent->left = x;
    }
    x->right = y;
    y->parent = x;
}

// Function to fix the tree after insertion
void fixInsertion(Node** root, Node* k) {
    Node* u;
    while (k->parent != NULL && k->parent->color == 0) {
        if (k->parent == k->parent->parent->right) {
            u = k->parent->parent->left;
            if (u->color == 0) {
                u->color = 1;
                k->parent->color = 1;
                k->parent->parent->color = 0;
                k = k->parent->parent;
            } else {
                if (k == k->parent->left) {
                    k = k->parent;
                    rightRotate(root, k);
                }
                k->parent->color = 1;
                k->parent->parent->color = 0;
                leftRotate(root, k->parent->parent);
            }
        } else {
            u = k->parent->parent->right;

            if (u->color == 0) {
                u->color = 1;
                k->parent->color = 1;
                k->parent->parent->color = 0;
                k = k->parent->parent;
            } else {
                if (k == k->parent->right) {
                    k = k->parent;
                    leftRotate(root, k);
                }
                k->parent->color = 1;
                k->parent->parent->color = 0;
                rightRotate(root, k->parent->parent);
            }
        }
        if (k == *root) {
            break;
        }
    }
    (*root)->color = 1;
}

// Function to insert a new node into the Red-Black Tree
void insertNode(Node** root, int data) {
    Node* newNode = createNode(data);
    if (*root == NULL) {
        *root = newNode;
    } else {
        Node* y = NULL;
        Node* x = *root;
        while (x != NULL) {
            y = x;
            if (newNode->data < x->data) {
                x = x->left;
            } else {
                x = x->right;
            }
        }
        newNode->parent = y;
        if (newNode->data < y->data) {
            y->left = newNode;
        } else {
            y->right = newNode;
        }
    }
    if (newNode->parent == NULL) {
        newNode->color = 1;
        return;
    }
    if (newNode->parent->parent == NULL) {
        return;
    }
    fixInsertion(root, newNode);
}

// Function to test the Red-Black Tree implementation
int main() {
    srand(time(NULL));
    Node* root = NULL;
    for (int i = 0; i < 1000; i++) {
        int data = rand() % 10000;
        insertNode(&root, data);
    }
    return 0;
}

File: Java/red-black tree/meta_llama_llama_4_maverick_17b_128e_instruct.java

import java.util.Random;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class meta_llama_llama_4_maverick_17b_128e_instruct {
  

    private static final boolean RED = true;
    private static final boolean BLACK = false;

    private static class Node {
        int key;
        Node left;
        Node right;
        Node parent;
        boolean color;

        public Node(int key, boolean color) {
            this.key = key;
            this.color = color;
            this.left = null;
            this.right = null;
            this.parent = null;
        }
    }

    private Node root;
    private Node TNULL;

    public meta_llama_llama_4_maverick_17b_128e_instruct() {
        TNULL = new Node(0, BLACK);
        root = TNULL;
    }

    private void preOrderHelper(Node node) {
        if (node != TNULL) {
            preOrderHelper(node.left);
            preOrderHelper(node.right);
        }
    }

    private void leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        if (y.left != TNULL) {
            y.left.parent = x;
        }
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }
        y.left = x;
        x.parent = y;
    }

    private void rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        if (y.right != TNULL) {
            y.right.parent = x;
        }
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.right) {
            x.parent.right = y;
        } else {
            x.parent.left = y;
        }
        y.right = x;
        x.parent = y;
    }

    private void fixInsert(Node k) {
        while (k.parent.color == RED) {
            if (k.parent == k.parent.parent.right) {
                Node u = k.parent.parent.left;
                if (u.color == RED) {
                    u.color = BLACK;
                    k.parent.color = BLACK;
                    k.parent.parent.color = RED;
                    k = k.parent.parent;
                } else {
                    if (k == k.parent.left) {
                        k = k.parent;
                        rightRotate(k);
                    }
                    k.parent.color = BLACK;
                    k.parent.parent.color = RED;
                    leftRotate(k.parent.parent);
                }
            } else {
                Node u = k.parent.parent.right;

                if (u.color == RED) {
                    u.color = BLACK;
                    k.parent.color = BLACK;
                    k.parent.parent.color = RED;
                    k = k.parent.parent;
                } else {
                    if (k == k.parent.right) {
                        k = k.parent;
                        leftRotate(k);
                    }
                    k.parent.color = BLACK;
                    k.parent.parent.color = RED;
                    rightRotate(k.parent.parent);
                }
            }
            if (k == root) {
                break;
            }
        }
        root.color = BLACK;
    }

    public void insert(int key) {
        Node node = new Node(key, RED);
        node.parent = null;
        node.left = TNULL;
        node.right = TNULL;

        Node y = null;
        Node x = root;

        while (x != TNULL) {
            y = x;
            if (node.key < x.key) {
                x = x.left;
            } else {
                x = x.right;
            }
        }

        node.parent = y;
        if (y == null) {
            root = node;
        } else if (node.key < y.key) {
            y.left = node;
        } else {
            y.right = node;
        }

        if (node.parent == null) {
            node.color = BLACK;
            return;
        }

        if (node.parent.parent == null) {
            return;
        }

        fixInsert(node);
    }


    public static void main(String[] args) {
        meta_llama_llama_4_maverick_17b_128e_instruct rbt = new meta_llama_llama_4_maverick_17b_128e_instruct();

        try (BufferedReader br = new BufferedReader(new FileReader("../../inputs.txt"))) {
            String line;
            while ((line = br.readLine()) != null) {
                int value = Integer.parseInt(line.trim());
                rbt.insert(value);
            }
        } catch (IOException e) {
            return;
        }
    }
}
